### PR TITLE
feat: vModel mounting option

### PR DESCRIPTION
**Pull request by lmiller1990**
Add a `vModel` mounting option.

I'd like lots of feedback. Please scrutinize - once we add this feature, we cannot go back or make any changes. We will need to maintain this for years to come. I am sure I am missing some tests and edge cases.

Also would like the opinion of @nekosaur and @kaelWD and @nandi95 - you are heavy users of test utils and have provided lots of useful feedback and contributions. Does this help with your `vModel` needs?

Component:
```vue
<template>
  <input @input="updateValue" />
  {{ color }}
</template>

<script>
export default {
  props: {
    color: String
  },
  methods: {
    handle($event) {
      this.$emit('update:color', $event.target.value)
    }
  }
}
</script>
```

Test:

```js
it('updates v-model automatically', async () => {
  const wrapper = mount(Component, {
    vModel: {
      color: 'red'
    }
  })

  expect(wrapper.html()).toContain('red')

  // update the input value
  wrapper.find('input').element.value = 'blue'
  // trigger the input event
  await wrapper.find('input').trigger('input')

  // value is updated and reflected in the DOM
  expect(wrapper.html()).toContain('blue')
})
```

resolves #https://github.com/vuejs/vue-test-utils-next/issues/321 and https://github.com/vuejs/vue-test-utils-next/discussions/279.

### CODE REVIEW

File: docs/api/index.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ interface MountingOptions<Props, Data = {}> {
   attrs?: Record<string, unknown>
   data?: () => {} extends Data ? any : Data extends object ? Partial<Data> : any
   props?: (RawProps & Props) | ({} extends Props ? null : never)
+  vModel?: (RawProps & Props) | ({} extends Props ? null : never)
```
Comment (pikax) [inline docs/api/index.md:19]: In terms of typing, should we just extract the props that the user has defined? eg: `update:${Prop}`

This is because vue-3 already gives a warning when using an non declared event.

File: docs/api/index.md [hunk 2]
```diff
@@ -244,6 +245,62 @@ test('props', () => {
...
+    }
+  }
+}
+</script>
+```
+
+`Component.spec.js`:
+
+```js
+it('updates v-model automatically', async () => {
```
Comment (afontcu) [inline docs/api/index.md:285]: Not that important, but examples in docs use `test` instead of `it` :) #nitpicking

```suggestion
test('updates v-model automatically', async () => {
```

File: docs/api/index.md [hunk 3]
```diff
@@ -244,6 +245,62 @@ test('props', () => {
...
+<template>
+  <input @input="updateValue" />
+  {{ color }}
+</template>
+
+<script>
+export default {
+  props: {
+    color: String
+  },
```
Comment (nandi95) [inline docs/api/index.md:272]: Perhaps use modelValue as example because I imagine `v-model=""` will be the most frequent use-case.
```suggestion
  },
  emits: ['update:modelValue'],
```

File: docs/api/index.md [hunk 4]
```diff
@@ -244,6 +245,62 @@ test('props', () => {
...
+      color: 'red'
+    }
+  })
+
+  expect(wrapper.html()).toContain('red')
+
+  // update the input value
+  wrapper.find('input').element.value = 'blue'
+  // trigger the input event
+  await wrapper.find('input').trigger('input')
```
Comment (afontcu) [inline docs/api/index.md:297]: I might be missing something here, but I would've expected something along the lines of:

```suggestion
  await wrapper.find('input').setValue('blue')
```

File: src/types.ts [hunk 5]
```diff
@@ -54,9 +54,14 @@ export interface MountingOptions<Props, Data = {}> {
    */
   props?: (RawProps & Props) | ({} extends Props ? null : never)
   /**
-   * @deprecated use `data` instead.
+   * @deprecated use `props` instead.
```
Comment (afontcu) [inline src/types.ts:57]: nice catch 👍

File: tests/mountingOptions/vModel.spec.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import { defineComponent, h } from 'vue'
+
+describe('vModel', () => {
+  it('updates v-model automatically', async () => {
+    const Comp = defineComponent({
+      props: ['foo'],
+      template: `
+        <input @input="handle" />
+        {{ foo }}
+      `,
```
Comment (nandi95) [inline tests/mountingOptions/vModel.spec.ts:13]: missing `emits: ['update:foo']` here, I believe

File: src/emit.ts [hunk 7]
```diff
@@ -64,4 +65,19 @@ export const recordEvent = (
 
   // Record the event message sent by the emit
   events[cid][event].push(args)
+
+  // see if we are emitting a v-model syntax sugar update:prop event
+  const match = event.match(/^(update:)(.*)/)
+
+  if (match && match[1] === 'update:' && vModelProps.includes(match[2])) {
+    if (args.length !== 1) {
+      throw new Error(
```
Comment (afontcu) [inline src/emit.ts:74]: Missing a test for this scenario? Not that common, but it would be great to know we don't run into some regression

File: tests/mountingOptions/vModel.spec.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        foo: 'foo'
+      }
+    })
+
+    expect(wrapper.html()).toContain('foo')
+    wrapper.find('input').element.value = 'bar'
+    await wrapper.find('input').trigger('input')
+    expect(wrapper.html()).toContain('bar')
+  })
+})
```
Comment (nandi95) [inline tests/mountingOptions/vModel.spec.ts:32]: Added missing test as pointed out by @afontcu 😂 
```suggestion
import { mount } from '../../src'
import { defineComponent, h } from 'vue'

describe('vModel', () => {
  const Comp = defineComponent({
    props: ['foo'],
    template: `
        <input @input="handle" />
        {{ foo }}
      `,
    emits: ['update:foo'],
    methods: {
      handle($event: KeyboardEvent) {
        this.$emit('update:foo', ($event.target as HTMLInputElement).value)
      }
    }
  })

  it('updates v-model automatically', async () => {
    const wrapper = mount(Comp, {
      vModel: {
        foo: 'foo'
      }
    })

    expect(wrapper.html()).toContain('foo')
    wrapper.find('input').element.value = 'bar'
    await wrapper.find('input').trigger('input')
    expect(wrapper.html()).toContain('bar')
  })

  it('should throw an error if payload is not 1 item', () => {
    const wrapper = mount(Comp, {
      vModel: {
        foo: 'foo'
      }
    })

    expect(() => wrapper.vm.$emit('update:foo')).toThrow(
      new Error('Two-way bound properties have to emit a single value. 0 values given.')
    )
    expect(() => wrapper.vm.$emit('update:foo', 1, 2)).toThrow(
      new Error('Two-way bound properties have to emit a single value. 2 values given.')
    )
  })
})
```

File: docs/api/index.md [hunk 9]
```diff
@@ -244,6 +245,62 @@ test('props', () => {
...
+vModel?: (RawProps & Props) | ({} extends Props ? null : never)
+```
+
+**Details:**
+
+`Component.vue`:
+
+```vue
+<template>
+  <input @input="updateValue" />
```
Comment (cexbrayat) [inline docs/api/index.md:264]: Should that call the `handle` method?

File: src/emit.ts [hunk 10]
```diff
@@ -64,4 +65,19 @@ export const recordEvent = (
 
   // Record the event message sent by the emit
   events[cid][event].push(args)
+
+  // see if we are emitting a v-model syntax sugar update:prop event
```
Comment (cexbrayat) [inline src/emit.ts:69]: nit: maybe add a comment to explain that, if that's the case, we want to update props automatically.

**Review by cexbrayat — COMMENTED** [review]
Code looks good, just a few nits.

I'm not sure I like the idea to add another option to the API just to save a bit of boilerplate, but if the team likes it, I don't mind.

**Comment by nandi95** [comment]
I could get used to having a `vModel` option. I think it saves a lot of writing if I don't have to write a wrapping component in the `mount` method [like this](https://github.com/vuejs/vue-test-utils-next/discussions/279#discussioncomment-432686)
I can't think of edge cases but regardless there is still time to find any between the release of this and a stable major release I would have thought.

**Comment by nekosaur** [comment]
@nandi95 that is one way of doing it, but a shorter one (with the exception of this PR) is simply

```js
const wrapper = mount(Component, {
  props: {
    modelValue: 5,
    'onUpdate:modelValue': modelValue => wrapper.setProps({ modelValue }),
  },
})
```

I'll try to find some time after work to have a look at this 👍

**Comment by pikax** [comment]
Other API could provide an vmodel wrapper:
```ts
import { vmodel, mount } from 'vue-test-utils'

mount(Component, {
 props: {
   modelValue: vmodel(5)
  }
}}
 
```

**Comment by lmiller1990** [comment]
I am also not 100% sold on having `vModel`. I don't really have a super strong feeling about it, though. This problem does come up often enough to justify having *something* (whether it's just better documentation, the `vModel` option, etc).

I'll go over the feedback closely and think about it more. No rush on merging this, happen to have alternative implementations explored!

**Comment by nandi95** [comment]
> @nandi95 that is one way of doing it, but a shorter one (with the exception of this PR) is simply
> 
> ```js
> const wrapper = mount(Component, {
>   props: {
>     modelValue: 5,
>     'onUpdate:modelValue': modelValue => wrapper.setProps({ modelValue }),
>   },
> })
> ```
> 
> I'll try to find some time after work to have a look at this 👍

@nekosaur I have found an issue with this approach.
I've got a component that has a v-model which could be `null`, `Record<string, unknown>`, `Record<string, unknown>[]`
This component has a prop called `multi`, which is a boolean indicating that multiple things can be selected resulting in `Record<string, unknown>[]`
The component has a watcher that runs immediately and if `multi` is true, emit the `modelValue` to be `[]` if the `modelValue` is `null` or wrap it in an array if not an array already.

When setting a listener in the props, it will not update the props from the immediate watcher.

**Comment by lmiller1990** [comment]
Good point @nandi95. Looking over VTU v1 there is many issues relating to watch and `setData` and `setProps`, for example: https://github.com/vuejs/vue-test-utils/issues/1756

I think this is one of the many complexities we should be aware of and why I'm very cautious about this new features.

@nandi95 does the wrapping method work in your use case?

**Comment by nandi95** [comment]
@lmiller1990 The wrapping the component works as expected.
However I'm against this approach as I have might voiced before because it requires a lot more writing which adds up when writing a lot of components.
Here's how it was possible to do in rc-4 vs rc-5(with wrapping):
Screenshot 2021-04-21 at 09 51 48

Note: that `nextTick` was needed for some reason to wait for the `watch` to update the `modelValue` before clicking.

**Comment by nekosaur** [comment]
@nandi95 @lmiller1990 The `vModel` option added in this PR does seem to support `watch` with `immediate: true`, since it will manually set the new prop value directly on `vm`.

It would be nice to be able to support `immediate: true` when doing it manually, but it looks like a catch 22

```ts
const wrapper = mount(Component as any, {
  props: {
    modelValue: null,
    'onUpdate:modelValue': (modelValue: any) => {
	  // With `immediate: true` this gets called before mounting is complete, which means wrapper is not yet assigned
      wrapper.setProps({ modelValue })
    }
  }
})
```

**Comment by lmiller1990** [comment]
I think passing a `onUpdate:modelValue` prop with a `setProps` combo is a little weird. I personally don't want to write that every time I write a test. I feel like a wrapper is a bit better (`setProps` feels like a bit of a hack to me, as opposed to wrapping with another component).

I think we should add a bunch more tests for complex cases to at least explore `vModel` as a mounting option. Things like `immediate: true` with `watch`, `watchEffect`, etc, and see if there are any edge cases before we serious consider actually merging this. 

Does anyone feel strongly about the `vModel` mounting option? I personally don't - I have a similar problem in React, I just write a wrapper, the extra code doesn't bother me all that much.

**Comment by lmiller1990** [comment]
This is still on the table, but not pre 2.0.0. I'm going to close it. We can revisit this once we hit a 2.0.0 release and Vue 3 is the main version on npm.

**Comment by nandi95** [comment]
For the time being if someone's willing, docs can be added which outlines how it can currently be tested
 - Passing in the listener suggestion by @nekosaur which works to some extent
 - Wrapping the component suggestion by @KaelWD

This later can be updated along the lines of the v-model PRs

**Comment by lmiller1990** [comment]
I will try to document something in the near future.
